ubi_manifest/app/api.py

```diff
@@ -1,21 +1,120 @@
-    return {"status": "OK"}
+@router.get(
+    "/status",
+    response_model=StatusResult,
+    responses={
+        200: {
+            "content": {
+                "application/json": {
+                    "example": {
+                        "server_status": "OK",
+                        "workers_status": {
+                            "availability": {"worker01": {"ok": "pong"}},
+                            "stats": {"worker01": {"some": "stats"}},
+                            "registered_tasks": {"worker01": ["some_task"]},
+                            "active_tasks": {"worker01": []},
+                            "scheduled_tasks": {"worker01": []},
+                        },
+                        "redis_status": {"Connected": True},
+                        "celery_beat_status": "OK",
+                        "connection_to_gitlab": "OK",
+                        "connection_to_pulp": "OK",
+                    }
+                }
+            },
+        }
+    },
+)
+def status() -> StatusResult:
+    # Check workers
+    i = app.control.inspect()
+    workers_status = {
+        "availability": i.ping(),
+        "stats": i.stats(),
+        "registered_tasks": i.registered(),
+        "active_tasks": i.active(),
+        "scheduled_tasks": i.scheduled(),
+    }
+
+    # Check redis
+    redis_client = redis.from_url(app.conf.result_backend)
+    redis_status = {"Connected": redis_client.ping()}
+
+    # Check celery beat
+    heartbeat = redis_client.get("celery-beat-heartbeat")
+
+    if heartbeat:
+        last_heartbeat = datetime.fromisoformat(heartbeat.decode("utf-8"))
+        now = datetime.now()
+        time_from_last_heartbeat = now - last_heartbeat
+        if time_from_last_heartbeat < timedelta(minutes=2):
+            celery_beat_status = {"status": "OK", "msg": "Celery beat operable."}
+        else:
+            days = time_from_last_heartbeat.days
+            hours = time_from_last_heartbeat.seconds // 3600
+            minutes = (time_from_last_heartbeat.seconds - hours * 3600) // 60
+            celery_beat_status = {
+                "status": "Failed",
+                "msg": f"Last heartbeat task ran {days} days, {hours} hours "
+                f"and {minutes} minutes ago.",
+            }
+    else:
+        celery_beat_status = {
+            "status": "n/a",
+            "msg": "No heartbeat task ran yet. Wait a minute.",
+        }
+
+    # Gitlab is used only if the content configs are loaded from Gitlab.
+    # In case the configs are loaded from directory, connection to Gitlab
+    # is not needed, therefore not checked.
+    gitlab_base_url = get_gitlab_base_url(app.conf.content_config)
+    if gitlab_base_url:
+        gitlab_resp = requests.get(
+            f"{gitlab_base_url}/-/health", timeout=REQUEST_TIMEOUT
+        )
+        gitlab_status = gitlab_resp.reason
+    else:
+        gitlab_status = "not needed"
+
+    # Check connection to Pulp
+    pulp_resp = requests.get(
```

Comment:
I think we should catch exceptions here  - this can raise e.g. `ConnectionError` and other exceptions as well (like TimeoutError) and then whole API call to ubi-manifest status endpoint will fail. We should try to catch most of exceptions if not all of them. 
Also I'm thinking about adding some retries as well - but I guess I'd wait for this to see how it behaves with real systems.

Reply:
Okay I guess I don't understand still. I thought you do NOT want to raise the exception so the whole status endpoint would NOT fail if pulp or gitlab fails (for whatever reason). I thought we want the status endpoint to successfully return their status, even though they are failing.
You said 
> I think we should catch all exceptions raised from this call and set the status as n/a or failed.

But what is the point in setting the gitlab or pulp status to anything, if you want to raise the exception, which would prevent the endpoint to return that status but would fail with respective error and user wouldn't know where is the problem unless they check the logs from worker where it is visible what exactly was the problem.
So do we want the whole endpoint to fail or to return successfully even with some failed statuses? 
Or am I still missing something here?

Reply:
This endpoint should return 200 with all available statuses. It doesn't matter if some dependency is not OK, it's up to client what they will do with returned data.

So any check for status you do in this endpoint should be robust enough so it will never fail whole endpoint.